File: Chapter 02/Python/section_2_3_3_testability/before/real_database_connection.py
"""
This is the "real" or "production" implementation of the database connection.
It would contain the actual logic to connect to and query a live database.

ARCHITECTURAL NOTE: In this "Before" state, notice that this class does not 
implement any abstract base class or interface. It is a rigid, concrete implementation.
"""

class RealDatabaseConnection:
    def __init__(self, connection_string: str):
        """
        Initializes the real database connection.
        """
        self._connection_string = connection_string
        # In a real application, this is where the connection would be established.
        print(f"\n  [DB] Connecting to... {self._connection_string}")

    def get_data(self, query: str) -> list[str]:
        """
        Fetches data from the live database.
        """
        # For demonstration purposes, we're just returning hardcoded data.
        print(f"  [DB] Executing query: {query}")
        return ["real_data_row1", "real_data_row2"]

File: Chapter 02/JS/section_2_3_3_testability/before/realDatabaseConnection.js
/**
 * This is the "real" or "production" implementation of the database connection.
 * It would contain the actual logic to connect to and query a live database.
 * In a real-world application, this class would use a library like Sequelize, 
 * Knex, or Mongoose to interact with a SQL or NoSQL database.
 * * ARCHITECTURAL NOTE: In this "Before" state, notice that this class does not 
 * implement any interface. It is a rigid, concrete implementation.
 */
class RealDatabaseConnection {
    #connectionString;

    /**
     * Initializes the real database connection.
     * @param {string} connectionString 
     */
    constructor(connectionString) {
        this.#connectionString = connectionString;
        // In a real application, this is where the connection would be established.
        console.log(`\n  [DB] Connecting to... ${this.#connectionString}`);
    }

    /**
     * Fetches data from the live database.
     * @param {string} query The query to execute.
     * @returns {string[]} A list of data rows from the real database.
     */
    getData(query) {
        // For demonstration purposes, we're just returning hardcoded data.
        // A real implementation would execute the query against the database
        // and return the results.
        console.log(`  [DB] Executing query: ${query}`);
        return ["real_data_row1", "real_data_row2"];
    }
}

module.exports = RealDatabaseConnection;